[pyqgis-console] using gist to share snippets instead of codepad

slarosa opened the pull request:
## Description
Refs #37574: replace codepad with gist to share snippet code. To create gist it is necessary to generate a personal access token from developer settings on the github profile settings. If token is set in general console settings the share menu will be enabled.

**Note**: _codepad.org is not working as it now is protected by reCAPTCHA._

![image](https://user-images.githubusercontent.com/1374682/93211351-66fd8300-f761-11ea-8411-4e6eb3dade98.png)
![image](https://user-images.githubusercontent.com/1374682/94159327-b475a080-fe83-11ea-9338-2f0d568535eb.png)

<!--
  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

slarosa committed:
[pyqgis-console] using gist to share snippets instead of codepad

elpaso commented:
How about using https://termbin.com/ ? No tokens or registration required.

slarosa commented:
> How about using https://termbin.com/ ? No tokens or registration required.

I was taking a look at termibin.com project and it seems a bit outdated (last commit was in December 2018 and last stable release on 30 Oct 2017) and I cannot understand if it has a way to create python snippets via public APIs. I only see example usage via terminal. Or maybe I'm looking in the wrong place (https://github.com/solusipse/fiche)??

Anyway, personally, I would prefer Gist platform for code snippets but we wait for some more comments on this.

elpaso commented:
I would recommend anything that does not require registration.

Something like this will work:

```python

from time import sleep
import socket

def netcat(hostname, port, content):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((hostname, port))
    s.sendall(content)
    sleep(0.5) 
    s.shutdown(socket.SHUT_WR)
    while 1:
        data = s.recv(1024)
        if data == b"":
            break
        print ("Received:", repr(data))
    print ("Connection closed.")
    s.close()

netcat('termbin.com', 9999, b'QGIS is nice and I really like anonymous services')
```

slarosa commented:
> I would recommend anything that does not require registration.

...and true, it is necessary to have a GitHub account to get a personal access token, but an advantage here is that doing so we encourage users to write more bug reports on GitHub :stuck_out_tongue_winking_eye: :smiley:

nyalldawson commented on python/console/console_editor.py:
```diff
@@ -358,36 +364,34 @@ def objectListEditor(self):
-        except urllib.error.URLError as e:
-            msgText = QCoreApplication.translate('PythonConsole', 'Connection error: ')
-            self.parent.pc.callWidgetMessageBarEditor(msgText + repr(e.args), 0, True)
+            res = requests.post(URL, headers=headers, params=params, data=json.dumps(data)).json()
```
Could this be changed to use QNetworkRequest or QgsBlockingNetworkRequest? Then we'd get proper handling of proxies, logging of the request, etc

nyalldawson replied:
(I can do that if you don't want to!)

slarosa replied:
I can give it a try, thanks for the insight!

slarosa committed:
[pyqgis-console] use QgsBlockingNetworkRequest to create Gist

slarosa commented:
@nyalldawson I made the changes to use QgsBlockingNetworkRequest to create the Gist.

nyalldawson commented:
Thanks! Looks great

rduivenvoorde commented:
@slarosa I tested your PR. Thanks, works, very cool! 
Some comments:
- the item 'share on GitHub' is not active in the menu unless you created a token. But there is no indication that you have to. Would it not be better to make it just always active, but throw a warning or throw the settings in the users face when they try to use it without having a token?
- about the token, we should really make it clear in the docs (or a link somewhere) how that would work https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token
- maybe instead of <ACCESS_TOKEN> use <github_personal_access_token> (as caption in the input in the settings) so users can google for it ( like I did :-) )?
- I see that the python console actually has a documentation link itself: https://docs.qgis.org/testing/en/docs/user_manual/plugins/python_console.html 
- What about adding a help button here in the settings dialog so the user can read about the token and such there? We do that in a lot of dialogs nowadays (see for example 'Add Vector Layer'-dialog in which the Help button brings you immidiatly to: https://docs.qgis.org/testing/en/docs/user_manual/managing_data_source/opening_data.html#loading-a-layer-from-a-file
- All works if I create a token only for GIST:

![Screenshot-20200917130354-680x1078](https://user-images.githubusercontent.com/731673/93462564-d05ace80-f8e6-11ea-8c40-ce6c9011a965.png)

rduivenvoorde commented:
@slarosa one other thing I want to add: would/could this functionality actually not be available in all editors (so also in 'console', 'output')... Or is that overkill?

With the new QgsPythonConsoleBase this is now super easy ;-)

slarosa commented:
> @slarosa one other thing I want to add: would/could this functionality actually not be available in all editors (so also in 'console', 'output')... Or is that overkill?
> 
> With the new QgsPythonConsoleBase this is now super easy ;-)

What do you mean by "not be available in all editors"? maybe you are saying to add it to the QgsPythonConsoleBase class?

rduivenvoorde commented:
> > @slarosa one other thing I want to add: would/could this functionality actually not be available in all editors (so also in 'console', 'output')... Or is that overkill?
> > With the new QgsPythonConsoleBase this is now super easy ;-)
> 
> What do you mean by "not be available in all editors"? maybe you are saying to add it to the QgsPythonConsoleBase class?

With 'editors' I mean to add the context menu to the three 'editor' classes: console_sci.py (for the actual command console) console_output.py (the output window of the console); and it is with this PR off course already available in console_editor.py.

slarosa commented:
> With 'editors' I mean to add the context menu to the three 'editor' classes: console_sci.py (for the actual command console) console_output.py (the output window of the console); and it is with this PR off course already available in console_editor.py.

sorry, ok now I get it, My concerns are that the output pane could be very dirty (traceback etc) and the command pane accepts only one line of code (just noticed). Editor pane instead may contain complete scripts or snippets which fit better eventually to share as Gist.
So I really would not know if that is useful to add the entry for each pane.

slarosa committed:
[pyqgis-console] adress some comment: add button to link at generate token page

slarosa commented:
added button to link at page to generate github token. (screenshot updated).
@rduivenvoorde I changed from ACCESS_TOKEN to PERSONAL_ACCESS_TOKEN, is it ok for you?

rduivenvoorde commented:
@slarosa this works very good, thanks!

After the merge I will try to add a Help button to the docs in the Options dialog.

Could we think of some 'indication' why 'Share on Github' is not yet available when there is no Token yet? 
Some idea's ... :
- make 'Share on Github' always enabled, BUT open the options dialog when there is no Token yet
or
- if there is no token keep the menu 'disabled' (like now), but make the label "Share on Github (See options)" ?

But I'm fine to merge first so the issue can be closed.

Thanks!

PS (nothing to do with this PR) but for the first time I noticed the big 'drop down button' all to the right to show the list of tabs:

![Screenshot-20200924212304-377x110](https://user-images.githubusercontent.com/731673/94190245-23ff8600-feac-11ea-9c5c-7803f049b124.png)

Isn't that a little big?

slarosa commented:
@rduivenvoorde anything like this?? with the entry menu always enabled ??
![image](https://user-images.githubusercontent.com/1374682/94256255-d32e7280-ff29-11ea-935b-464442494fa2.png)

rduivenvoorde commented:
Yes! Looks good to me! Thanks!

slarosa committed:
[pyqgis-console] warn user to generate personal access token

nyalldawson commented:
@slarosa that messagebar warning should be a warning or critical level message, not info level. 

Reading over this PR it seems like all other requested changes have been addressed. @rduivenvoorde are you happy for it to be merged now?

slarosa committed:
[pyqgis-console] change message bar to Warning

rduivenvoorde commented:
Yes yes, Please merge :-)

NathanW2 commented:
Nice work

On Sun, 27 Sep 2020, 4:55 am qgis-bot, <notifications@github.com> wrote:

> @slarosa <https://github.com/slarosa>
> A documentation ticket has been opened at qgis/QGIS-Documentation#6131
> <https://github.com/qgis/QGIS-Documentation/issues/6131>
> It is *your* responsibility to visit this ticket and add as much detail
> as possible for the documentation team to correctly document this change.
> Thank you!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/38768#issuecomment-699534517>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC5FXFXZYETNJU3OGWVYC3SHY2LDANCNFSM4RNFRO2A>
> .
>

slarosa commented:
Thanks :-)
